src/logic/div.rs

```diff
@@ -0,0 +1,1290 @@
+                false
+            },
+        };
+        ApInt::wrapping_udivrem_assign(self, &mut rhs_clone).unwrap();
+        if negate_lhs {self.wrapping_neg()}
+        //clearing unused bits is handled by `wrapping_neg()`
+        Ok(())
+    }
+
+    /// Divides `self` by `rhs` using **signed** interpretation and returns the quotient.
+    pub fn into_wrapping_sdiv(self, rhs: &ApInt) -> Result<ApInt> {
+        try_forward_bin_mut_impl(self, rhs, ApInt::wrapping_sdiv_assign)
+    }
+
+    /// Remainder-assigns `lhs` by `rhs` inplace using **signed** interpretation.
+    pub fn wrapping_srem_assign(&mut self, rhs: &ApInt) -> Result<()> {
+        if rhs.is_zero() {
+            return Err(Error::division_by_zero(DivOp::SignedRem, self.clone()))
+        }
+        let mut rhs_clone = (*rhs).clone();
+        let negate_lhs = match ((*self).is_negative(), rhs_clone.is_negative()) {
+            (false,false) => false,
+            (true,false) => {
+                self.wrapping_neg();
+                true
+            },
+            (false,true) => {
+                rhs_clone.wrapping_neg();
+                false
+            },
+            (true,true) => {
+                self.wrapping_neg();
+                rhs_clone.wrapping_neg();
+                true
+            },
+        };
+        ApInt::wrapping_uremdiv_assign(self, &mut rhs_clone).unwrap();
+        if negate_lhs {self.wrapping_neg()}
+        //clearing unused bits is handled by `wrapping_neg()`
+        Ok(())
+    }
+
+    /// Divides `self` by `rhs` using **signed** interpretation and returns the remainder.
+    pub fn into_wrapping_srem(self, rhs: &ApInt) -> Result<ApInt> {
+        try_forward_bin_mut_impl(self, rhs, ApInt::wrapping_srem_assign)
+    }
+}
+
+#[cfg(test)]
+#[rustfmt::skip]
+mod tests {
+    use super::*;
+    use crate::info::BitWidth;
+
+    mod div_rem {
+        use super::*;
+        use std::u64;
+
+        //TODO: add division by zero testing after error refactoring is finished
+        //use errors::ErrorKind;
+        #[test]
+        fn simple() {
+            /// does all of the simple division tests
+            /// - `$signed`: if the functions are signed divisions or not
+            /// - `$fun_assign`: a division function such as `wrapping_udiv_assign` with that
+            ///     signature
+            /// - `$fun_into`: a division function such as `into_wrapping_udiv` with that signature
+            /// - `$fun`: a division function such as `wrapping_udivrem_assign` with that signature
+            /// - `$r0`: the quotient or remainder or both of 80 by 7, depending on division
+            ///     function type
+            /// - `$r1`, `$r2`, `$r3`: 80 by -7, -80 by 7, -80 by -7. These can be 0 if `$signed` is
+            ///     false.
+            macro_rules! s {
+                ($signed:expr,$fun_assign:ident,$fun_into:ident,$r0:expr,$r1:expr,$r2:expr,$r3:expr/*,$div_op:ident*/) => {
```

Comment:
please make use of the newline character to make this at least readable in theory.